Implement new UI for contributor admin dashboard Desktop+Mobile: GSoC M1.7

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #17440  .
2. This PR does the following: Implements new UI for contributor-admin-dashboard, along with feature flag `CD_NEW_UI`. It implements both Desktop and Mobile views.
3. Link to the mocks : [Mocks](https://www.figma.com/file/bmyOkQwosePcoNAr45Rdk0/Admin-Dashboard?node-id=35%3A3752&mode=dev)

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network
[Screencast from 26-07-23 09:33:23 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/41394eb9-74de-4276-94b0-67408fd88920)



<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone
[Screencast from 26-07-23 09:35:29 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/ad5480fa-8bc9-437b-87ac-b2ab72233a43)



<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
added feature flag for new cd admin

kshitij01042002 committed:
resolved merge comflicts

kshitij01042002 committed:
temporary changes

kshitij01042002 committed:
temporary commit

kshitij01042002 committed:
initial commit

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc_M1.7

kshitij01042002 committed:
temp commit

kshitij01042002 committed:
temp changes

kshitij01042002 committed:
temp commit

kshitij01042002 committed:
temp commit

kshitij01042002 committed:
merged with develop

kshitij01042002 committed:
implemented desktop and mobile view

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.7

qinghaoyang commented:
@kshitij01042002 Under Question Submitters, `1, (0 without edits)` seems not right. Do you want to show `1 (0 without edits)`?

kshitij01042002 commented:
- [x]      The width of the two tables on the page should match each other.
- [x] The info tooltips should go after their headers, not before.
- [x] There isn't a demo of how the the info tooltips get used -- please show for both desktop and mobile. Note that hovering is not a possible action on mobile devices. - Thanks for pointing out, added the feature that the tooltip will be enabled on clicking on the info icon.
- [x] When switching tabs, keep "User" in the same position.
- [x] Have a separate video for mobile view and put it in the "mobile" section.
- [x] I don't think you need to capitalize "Cards" and "Words" in the translator user expansion.(Did it according to mocks but have changed on your suggestion)
- [x] In the list of active topics, there should be a space after each comma.
- [x] In the question contributor expansion, there should not be a comma after "1". It should be "1 (0 without edits)".
- [x] In the mobile view, the table below should be the same width as the summary stats table at the top.
- [x] In mobile view, do not have a space before each of the colons.


Hi @seanlip I have changed the UI according to your suggestion, can you please take a another pass? Thanks for the feedback.

kshitij01042002 committed:
made changes on the first pass of UI

kshitij01042002 committed:
merge with upstream

seanlip commented:
I've sent some UI feedback. Make sure to get tests passing, too!

seanlip commented:
(Here it is, for posterity)

Desktop view:
- For "Translation Reviewers", you have a space between "card" and "s", and the "0 cards, words" next to "Rejected Translations" doesn't make sense.
- When you click on "Question Submitters", the table looks odd for a while before reverting to its proper state. That odd interlude should not happen. Is the table actively loading data from the backend at this point? If so, you should show a loading message.
- You still have "Edits" capitalized in some places. Please fix.
- There is a similar "flash of content" when loading the Question Reviewers tab; this should be fixed as well.
- The "Language/Question coordinators" tabs should be hidden/disabled until you have things to put in them.

Mobile view:
- I think you don't need so much spacing between the lines in the expandable card.
- I would suggest a 3-column layout for the top part of each row. In the first two columns, put what you already have, but move the "Edit Role" to the rightmost column since it's different from the others. That way, you have space to put a button that's large enough to capture the user's click (currently you have a rather small click area and it's easy to press the downward chevron icon by mistake).
- The stats at the top don't look nice. The numbers fall below the partitions set by the vertical lines; you might want to move them a bit higher (and possibly make the vertical lines slightly longer). Right now it looks messy, so try and use your aesthetic judgement to make it look nice.

Some general comments:
- I have a feeling your data descriptions are wrong. Please double-check to make sure, when you say "without edits", you're correctly referring to cards and not words. I think the format could be something like "X cards (Y without edits), Z words".
- In the detail cards, "submitted" should come before "accepted", which should come before "rejected".
- For the tooltip text, use the following:
  - Performance is calculated based on the review results of the last 100 accepted/rejected contributions.
  - Accuracy is based on the ratio of accepted to submitted translations.
- The next time you send a request for review, I recommend that you don't use fake data, but actually start loading stuff from the backend. This will help you work on the thing that needs to be shown while the data is still loading.

kshitij01042002 committed:
added tests

kshitij01042002 commented:
Hi @seanlip I have worked on your suggestion. PTAL. Also the data is coming from backend and not hard coded. 
Thanks for the feedbacks.

seanlip commented:
Thanks @kshitij01042002, this is much better. Only a few comments left:

- For translation submitters, I suggest "X cards (Y without edits), Z words" instead of what you have now -- i.e. drop the "cards" inside the parentheses. That makes it less crowded on mobile view.
- For translation reviewers, I suggest "X cards (Y edited), Z words".
- The part inside the parentheses is different for translation submitters and translation reviewers. I think that is fine actually (because it measures "perfect translations" by submitters, and work done by reviewers, respectively) but I wanted to ask that you double-check that you are pulling the right data to populate these fields, and that this isn't an accidental mislabeling.

Otherwise the UI looks great!

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM

kshitij01042002 committed:
added tests

kshitij01042002 committed:
final commit

kshitij01042002 committed:
eslint errors

kshitij01042002 committed:
lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
lint errors

kshitij01042002 committed:
identation errors

kshitij01042002 committed:
identation errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
ts errors

kshitij01042002 committed:
fixed ts checks

kshitij01042002 committed:
fixed ts checks

kshitij01042002 committed:
fixed ts checks

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.7

kshitij01042002 committed:
fixed failing tests

kshitij01042002 committed:
fixed failing tests

kshitij01042002 committed:
frontend coverage

kshitij01042002 committed:
Merge branch 'gsoc_M1.7' of https://github.com/kshitij01042002/oppia into gsoc_M1.7

kshitij01042002 committed:
removed fdescribe

kshitij01042002 committed:
removed commit value from constants

kshitij01042002 committed:
covered full coverage

Lawful2002 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -0,0 +1,62 @@
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for contributor dashboard page component.
```
```suggestion
 * @fileoverview Unit tests for contributor admin dashboard page component.
```

kshitij01042002 replied:
Changed.

Lawful2002 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.spec.ts:
```diff
@@ -0,0 +1,62 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
```suggestion
// Copyright 2023 The Oppia Authors. All Rights Reserved.
```

kshitij01042002 replied:
Changed.

kshitij01042002 committed:
failing backend test

seanlip commented:
@kshitij01042002 You didn't address some of my comments here (https://github.com/oppia/oppia/pull/18669#issuecomment-1649607805) or reply to the third one, so I am sending this back to you. 

Please respond to all comments and only then request a review. Thanks.

kshitij01042002 commented:
> @kshitij01042002 You didn't address some of my comments here ([#18669 (comment)](https://github.com/oppia/oppia/pull/18669#issuecomment-1649607805)) or reply to the third one, so I am sending this back to you.
> 
> Please respond to all comments and only then request a review. Thanks.

Ohhh Sorry! I implemented your suggestions but forgot to specifically address them. Also yes the values are populated with the right data, for reviewers we are storing values "with edits' and for submitters we are saving "without edits"

kshitij01042002 committed:
fixed failing backend test

seanlip commented:
No, I'm not just talking about responding to the comments. You have not implemented the wording changes I suggested in the comment I linked to (from looking at the video).

JayVivarekar reviewed (approved):
Thanks @kshitij01042002! LGTM for codeowner files.

kshitij01042002 commented:
> No, I'm not just talking about responding to the comments. You have not implemented the wording changes I suggested in the comment I linked to (from looking at the video).

Updated. PTAL. Thanks.

kshitij01042002 committed:
minor suggestion update

Lawful2002 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-dashboard-admin-summary.model.ts:
```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
```
```suggestion
// Copyright 2023 The Oppia Authors. All Rights Reserved.
```

kshitij01042002 replied:
Changed. Thanks for the nit.

Lawful2002 reviewed (changes requested):
Thanks @kshitij01042002, a few minor nits.

seanlip commented on core/domain/platform_parameter_list.py:
```diff
@@ -58,6 +58,7 @@ class ParamNames(enum.Enum):
         'high_bounce_rate_task_state_bounce_rate_obsoletion_threshold')
     HIGH_BOUNCE_RATE_TASK_MINIMUM_EXPLORATION_STARTS = (
         'high_bounce_rate_task_minimum_exploration_starts')
+    CD_NEW_UI = 'cd_new_ui'
```
This is too broad. CD_ADMIN_DASHBOARD_NEW_UI would be better, because the CD refers to /contributor-dashboard and we might need to do other things there. Make your feature flags specific.

kshitij01042002 replied:
Sound reasonable. Updated. PTAL.

seanlip commented on core/templates/domain/platform_feature/feature-status-summary.model.ts:
```diff
@@ -37,7 +37,8 @@ export enum FeatureNames {
   IsImprovementsTabEnabled = 'is_improvements_tab_enabled',
-  LearnerGroupsAreEnabled = 'learner_groups_are_enabled'
+  LearnerGroupsAreEnabled = 'learner_groups_are_enabled',
+  CD_NEW_UI = 'cd_new_ui'
```
This does not follow the existing UpperCamelCase convention.

kshitij01042002 replied:
Fixed.

seanlip reviewed (changes requested):
UI looks good, thanks. Just two code comments.

chris7716 commented on core/controllers/contributor_dashboard_admin_test.py:
```diff
@@ -739,10 +739,23 @@ def setUp(self) -> None:
         user_services.add_user_role(
             self.contributor_id, feconf.ROLE_ID_TRANSLATION_ADMIN)
 
+        user_settings_1 = user_services.create_new_user(
```
Can you rename these variables to have proper meanings?

kshitij01042002 replied:
Fixed. PTAL again.

chris7716 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-filter.model.ts:
```diff
@@ -0,0 +1,48 @@
+ */
+export class ContributorAdminDashboardFilter {
+  languageCode?: string;
+  topicIds?: string[];
```
Why topicIds optional?

kshitij01042002 replied:
Yeah, topicIds shouldn't be optional. Fixed.

chris7716 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-filter.model.ts:
```diff
@@ -0,0 +1,48 @@
+  languageCode?: string;
+  topicIds?: string[];
+  sort?: string;
+  lastActivity?: number;
```
Why is this optional?

kshitij01042002 replied:
These are supposed to be optional cause will making the request to the handler, we only have to pass filters which are set otherwise we are not supposed to pass them. This is partially implemented, it will make more sense once we implement filtering and sorting in frontend as well.

chris7716 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-filter.model.ts:
```diff
@@ -0,0 +1,48 @@
+ *   ContributorAdminDashboardFilter instance.
+ */
+  static createDefault(): ContributorAdminDashboardFilter {
+    return new ContributorAdminDashboardFilter('es', []);
```
Why `'es'`? Can it be a const?

kshitij01042002 replied:
This is passed as hardcoded for now, till we implement sorting and filtering.

chris7716 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-admin-dashboard-page.component.ts:
```diff
@@ -0,0 +1,84 @@
+  TAB_NAME_TRANSLATION_REVIEWER: string = 'Translation Reviewer';
+  TAB_NAME_QUESTION_SUBMITTER: string = 'Question Submitter';
+  TAB_NAME_QUESTION_REVIEWER: string = 'Question Reviewer';
+  translationReviewersCount: number = 24;
```
Why are these 24? Here and below. Can they be const somewhere?

kshitij01042002 replied:
Sorry, it was dummy values while testing. Fixed.

chris7716 commented on core/templates/pages/contributor-dashboard-admin-page/contributor-dashboard-tables/contributor-admin-stats-table.component.ts:
```diff
@@ -0,0 +1,244 @@
+      this.ContributorDashboardAdminStatsBackendApiService
+        .fetchContributorAdminStats(
+          ContributorAdminDashboardFilter.createDefault(),
+          20,
```
Why 20? Can it be a const?

kshitij01042002 replied:
It is defaulted for now, this will change according to user preference.All the pagination, sorting and filtering from frontend is hardcoded for now.

chris7716 reviewed (changes requested):
Thanks! Took a pass. @kshitij01042002

kshitij01042002 committed:
implemented suggestions by Harsh, Sean and Hasitha

kshitij01042002 committed:
fixed failing tests

kshitij01042002 commented:
Hi @Lawful2002 @chris7716 @seanlip I have addressed all your comments, PTAL again. Thanks for the review.

kshitij01042002 committed:
fixed failing e2e tests

kshitij01042002 committed:
feature flag definition error

kshitij01042002 committed:
fixed frontend test

seanlip reviewed (approved):
Thanks, LGTM for codeowners.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M1.7

Lawful2002 reviewed (approved):
Thanks, LGTM for frontend tests

chris7716 reviewed (approved):
Thanks! Approving since few of my suggestions are going to be implemented in a coming PR.
